File: python/hackerrank/simulation/flatlandSpaceStations.py

def flatlandSpaceStations(n, m, c):
    c.sort()
    max_dist = c[0]
    max_dist = max(max_dist, (n - 1) - c[-1])

    for i in range(1, m):
        max_dist = max(max_dist, (c[i] - c[i-1]) // 2)
    
    return max_dist
    
def main():
    n, m = map(int, input().split())
    c = list(map(int, input().rstrip().split()))
    print(flatlandSpaceStations(n, m, c))

if __name__ == "__main__":
    main()

File: golang/hackerrank/simulation/flatlandSpaceStations.go
package simulation

import (
	"sort"
)

/*
FlatlandSpaceStations calculates the maximum distance from any city to its nearest space station.

Args:

	n: The number of cities.
	m: The number of space stations.
	c: An integer slice of the indices of cities that have a space station.

Returns:

	int: The maximum distance any city is from a space station.
*/
func FlatlandSpaceStations(n int, m int, c []int) int {
	sort.Ints(c)

	maxDist := c[0]
	lastDist := (n - 1) - c[m-1]

	if lastDist > maxDist {
		maxDist = lastDist
	}

	for i := 1; i < m; i++ {
		dist := (c[i] - c[i-1]) / 2
		if dist > maxDist {
			maxDist = dist
		}
	}

	return maxDist
}
